File: modules/native/testing/integration/iam/identity/active.test.ts
import { randomBytes } from 'crypto'
import { ObjectId } from 'mongodb'
import { Status } from '@grpc/grpc-js/build/src/constants'

import { RequestError as GRPCRequestError } from '../../../../../system/libs/ts/grpc'
import { client as mongoClient, connect as connectToMongo, close as closeMongo } from '../../../../../system/testing/tools/mongo'
import { client as cacheClient, connect as connectToCache, close as closeCache } from '../../../../../system/testing/tools/cache'
import { client as nativeNamespaceGRPC, connect as connectToNativeNamespace, close as closeNativeNamespace } from '../../../tools/namespace/grpc'
import { identityClient as nativeIAmIdentityGRPC, connect as connectToNativeIAM, close as closeNativeIAM } from '../../../tools/iam/grpc'

const GLOBAL_DB_NAME = `${process.env.SYSTEM_DB_PREFIX || "openerp_"}global`
const TEST_NAMESPACE_NAME = "iamidentitytestnamespace"
const NAMESPACE_DB_NAME = `${process.env.SYSTEM_DB_PREFIX || "openerp_"}namespace_${TEST_NAMESPACE_NAME}`


beforeAll(async ()=>{
    await connectToMongo()
    await connectToCache()
    await connectToNativeIAM()
    await connectToNativeNamespace()
    
})

beforeEach(async () => {
    await nativeNamespaceGRPC.Ensure({ name: TEST_NAMESPACE_NAME })
})

afterEach(async ()=>{
    try {
        await mongoClient.db(GLOBAL_DB_NAME).collection('native_iam_identity').deleteMany({})
    } catch {}
    try {
        await mongoClient.db(NAMESPACE_DB_NAME).collection('native_iam_identity').deleteMany({})
    } catch {}
    await nativeNamespaceGRPC.Delete({ name: TEST_NAMESPACE_NAME })
    await cacheClient.flushall()
})

afterAll(async ()=>{
    await closeMongo()
    await closeCache()
    await closeNativeIAM()
    await closeNativeNamespace()
})

/**
 * @group native/iam/identity/active/whitebox
 * @group whitebox
 */
describe("Whitebox", () => {
    test("Updates identity in global database", async () => {
        const response = await nativeIAmIdentityGRPC.Create({
            name: "test",
            initiallyActive: false,
            namespace: ""
        })
        const id = ObjectId.createFromHexString(response.identity?.uuid as string)

        await nativeIAmIdentityGRPC.SetActive({
            namespace: "",
            uuid: id.toHexString(),
            active: true
        })

        const entry = await mongoClient.db(GLOBAL_DB_NAME).collection("native_iam_identity").findOne<{ active: boolean }>({"_id": id})
        expect(entry?.active).toBe(true)
    })

    test("Updates identity in namespace database", async () => {
        const response = await nativeIAmIdentityGRPC.Create({
            name: "test",
            initiallyActive: false,
            namespace: TEST_NAMESPACE_NAME
        })
        const id = ObjectId.createFromHexString(response.identity?.uuid as string)

        await nativeIAmIdentityGRPC.SetActive({
            namespace: TEST_NAMESPACE_NAME,
            uuid: id.toHexString(),
            active: true
        })

        const entry = await mongoClient.db(NAMESPACE_DB_NAME).collection("native_iam_identity").findOne<{ active: boolean }>({"_id": id})
        expect(entry?.active).toBe(true)
    })

    test("Clears cache on update", async () => {
        const response = await nativeIAmIdentityGRPC.Create({
            name: "test",
            initiallyActive: false,
            namespace: TEST_NAMESPACE_NAME
        })
        const id = ObjectId.createFromHexString(response.identity?.uuid as string)

        await nativeIAmIdentityGRPC.Get({ namespace: "", uuid: id.toHexString(), useCache: true })
        let existResponse = await cacheClient.exists(`native_iam_identity_data__${id.toHexString()}`)
        expect(existResponse).toBe(1)
        
        await nativeIAmIdentityGRPC.SetActive({
            namespace: TEST_NAMESPACE_NAME,
            uuid: id.toHexString(),
            active: true
        })

        existResponse = await cacheClient.exists(`native_iam_identity_data__${id.toHexString()}`)
        expect(existResponse).toBe(0)
    })
})

/**
 * @group native/iam/identity/active/blackbox
 * @group blackbox
 */
describe("Blackbox", () => {
    test("Returns new identity information as a result of update", async () => {
        const name = randomBytes(32).toString('hex')
        
        const createResponse = await nativeIAmIdentityGRPC.Create({
            name,
            initiallyActive: false,
            namespace: TEST_NAMESPACE_NAME
        })
        const id = ObjectId.createFromHexString(createResponse.identity?.uuid as string)

        const activeResponse = await nativeIAmIdentityGRPC.SetActive({
            namespace: TEST_NAMESPACE_NAME,
            uuid: id.toHexString(),
            active: true
        })

        expect(activeResponse.identity?.uuid).toBe(id.toHexString())
        expect(activeResponse.identity?.namespace).toBe(TEST_NAMESPACE_NAME)
        expect(activeResponse.identity?.active).toBe(true)
        expect(activeResponse.identity?.name).toBe(name)
        expect(activeResponse.identity?.policies).toStrictEqual([])
    })

    test("Get request will return new data after update", async () => {
        const createResponse = await nativeIAmIdentityGRPC.Create({
            name: "test",
            initiallyActive: false,
            namespace: TEST_NAMESPACE_NAME
        })
        const id = ObjectId.createFromHexString(createResponse.identity?.uuid as string)

        await nativeIAmIdentityGRPC.SetActive({
            namespace: TEST_NAMESPACE_NAME,
            uuid: id.toHexString(),
            active: true
        })

        const getResponse = await nativeIAmIdentityGRPC.Get({
            namespace: TEST_NAMESPACE_NAME,
            uuid: id.toHexString(),
            useCache: false
        })

        expect(getResponse.identity?.active).toBe(true)
    })

    test("Multiple updates resuts in same result", async () => {
        const createResponse = await nativeIAmIdentityGRPC.Create({
            name: "test",
            initiallyActive: false,
            namespace: TEST_NAMESPACE_NAME
        })
        const id = ObjectId.createFromHexString(createResponse.identity?.uuid as string)

        for (let i = 0; i < 5; i += 1) {
            await nativeIAmIdentityGRPC.SetActive({
                namespace: TEST_NAMESPACE_NAME,
                uuid: id.toHexString(),
                active: true
            })
            const getResponse = await nativeIAmIdentityGRPC.Get({
                namespace: TEST_NAMESPACE_NAME,
                uuid: id.toHexString(),
                useCache: false
            })
            expect(getResponse.identity?.active).toBe(true)
        }

        for (let i = 0; i < 5; i += 1) {
            await nativeIAmIdentityGRPC.SetActive({
                namespace: TEST_NAMESPACE_NAME,
                uuid: id.toHexString(),
                active: false
            })
            const getResponse = await nativeIAmIdentityGRPC.Get({
                namespace: TEST_NAMESPACE_NAME,
                uuid: id.toHexString(),
                useCache: false
            })
            expect(getResponse.identity?.active).toBe(false)
        }
    })

    test("Failes with INVALID_ARGUMENT if uuid has bad format", async () => {
        try {
            await nativeIAmIdentityGRPC.SetActive({
                namespace: TEST_NAMESPACE_NAME,
                uuid: "invalid",
                active: false
            })
            fail()
        } catch(e) {
            expect((e as GRPCRequestError)?.code).toBe(Status.INVALID_ARGUMENT)
        }
    })

    test("Failes with NOT_FOUND error if identity doesnt exist", async () => {
        try {
            await nativeIAmIdentityGRPC.SetActive({
                namespace: TEST_NAMESPACE_NAME,
                uuid: new ObjectId().toHexString(),
                active: false
            })
            fail()
        } catch(e) {
            expect((e as GRPCRequestError)?.code).toBe(Status.NOT_FOUND)
        }
    })
})

File: modules/native/testing/integration/iam/identity/active_test.go
package identity

import (
	"context"
	"testing"
	"time"

	"github.com/stretchr/testify/require"
	"github.com/stretchr/testify/suite"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"

	native "github.com/slamy-solutions/openbp/modules/native/libs/golang"
	"github.com/slamy-solutions/openbp/modules/native/libs/golang/iam/identity"
	"github.com/slamy-solutions/openbp/modules/native/libs/golang/namespace"
	"github.com/slamy-solutions/openbp/modules/native/testing/tools"
)

type ActiveTestSuite struct {
	suite.Suite

	nativeStub *native.NativeStub
}

func (suite *ActiveTestSuite) SetupSuite() {
	suite.nativeStub = native.NewNativeStub(native.NewStubConfig().WithNamespaceService().WithIAMIdentityService())
	err := suite.nativeStub.Connect()
	if err != nil {
		panic(err)
	}
}
func (suite *ActiveTestSuite) TearDownSuite() {
	suite.nativeStub.Close()
}
func TestActiveTestSuite(t *testing.T) {
	suite.Run(t, new(ActiveTestSuite))
}

func (s *ActiveTestSuite) TestInGlobalNamespace() {
	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
	defer cancel()

	createResponse, err := s.nativeStub.Services.IamIdentity.Create(ctx, &identity.CreateIdentityRequest{
		Namespace:       "",
		Name:            tools.GetRandomString(20),
		InitiallyActive: true,
		Managed:         &identity.CreateIdentityRequest_No{No: &identity.NotManagedData{}},
	})
	require.Nil(s.T(), err)
	defer s.nativeStub.Services.IamIdentity.Delete(context.Background(), &identity.DeleteIdentityRequest{Namespace: "", Uuid: createResponse.Identity.Uuid})

	_, err = s.nativeStub.Services.IamIdentity.SetActive(ctx, &identity.SetIdentityActiveRequest{
		Namespace: "",
		Uuid:      createResponse.Identity.Uuid,
		Active:    false,
	})
	require.Nil(s.T(), err)
	getResponse, err := s.nativeStub.Services.IamIdentity.Get(ctx, &identity.GetIdentityRequest{
		Namespace: "",
		Uuid:      createResponse.Identity.Uuid,
		UseCache:  false,
	})
	require.Nil(s.T(), err)
	require.False(s.T(), getResponse.Identity.Active)

	_, err = s.nativeStub.Services.IamIdentity.SetActive(ctx, &identity.SetIdentityActiveRequest{
		Namespace: "",
		Uuid:      createResponse.Identity.Uuid,
		Active:    true,
	})
	require.Nil(s.T(), err)
	getResponse, err = s.nativeStub.Services.IamIdentity.Get(ctx, &identity.GetIdentityRequest{
		Namespace: "",
		Uuid:      createResponse.Identity.Uuid,
		UseCache:  false,
	})
	require.Nil(s.T(), err)
	require.True(s.T(), getResponse.Identity.Active)

	_, err = s.nativeStub.Services.IamIdentity.SetActive(ctx, &identity.SetIdentityActiveRequest{
		Namespace: "",
		Uuid:      createResponse.Identity.Uuid,
		Active:    false,
	})
	require.Nil(s.T(), err)
	getResponse, err = s.nativeStub.Services.IamIdentity.Get(ctx, &identity.GetIdentityRequest{
		Namespace: "",
		Uuid:      createResponse.Identity.Uuid,
		UseCache:  false,
	})
	require.Nil(s.T(), err)
	require.False(s.T(), getResponse.Identity.Active)
}

func (s *ActiveTestSuite) TestInNamespace() {
	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
	defer cancel()

	namespaceName := tools.GetRandomString(20)
	_, err := s.nativeStub.Services.Namespace.Create(ctx, &namespace.CreateNamespaceRequest{
		Name:        namespaceName,
		FullName:    "",
		Description: "",
	})
	require.Nil(s.T(), err)
	defer s.nativeStub.Services.Namespace.Delete(context.Background(), &namespace.DeleteNamespaceRequest{Name: namespaceName})

	createResponse, err := s.nativeStub.Services.IamIdentity.Create(ctx, &identity.CreateIdentityRequest{
		Namespace:       namespaceName,
		Name:            tools.GetRandomString(20),
		InitiallyActive: true,
		Managed:         &identity.CreateIdentityRequest_No{No: &identity.NotManagedData{}},
	})
	require.Nil(s.T(), err)
	defer s.nativeStub.Services.IamIdentity.Delete(context.Background(), &identity.DeleteIdentityRequest{Namespace: namespaceName, Uuid: createResponse.Identity.Uuid})

	_, err = s.nativeStub.Services.IamIdentity.SetActive(ctx, &identity.SetIdentityActiveRequest{
		Namespace: namespaceName,
		Uuid:      createResponse.Identity.Uuid,
		Active:    false,
	})
	require.Nil(s.T(), err)
	getResponse, err := s.nativeStub.Services.IamIdentity.Get(ctx, &identity.GetIdentityRequest{
		Namespace: namespaceName,
		Uuid:      createResponse.Identity.Uuid,
		UseCache:  false,
	})
	require.Nil(s.T(), err)
	require.False(s.T(), getResponse.Identity.Active)

	_, err = s.nativeStub.Services.IamIdentity.SetActive(ctx, &identity.SetIdentityActiveRequest{
		Namespace: namespaceName,
		Uuid:      createResponse.Identity.Uuid,
		Active:    true,
	})
	require.Nil(s.T(), err)
	getResponse, err = s.nativeStub.Services.IamIdentity.Get(ctx, &identity.GetIdentityRequest{
		Namespace: namespaceName,
		Uuid:      createResponse.Identity.Uuid,
		UseCache:  false,
	})
	require.Nil(s.T(), err)
	require.True(s.T(), getResponse.Identity.Active)

	_, err = s.nativeStub.Services.IamIdentity.SetActive(ctx, &identity.SetIdentityActiveRequest{
		Namespace: namespaceName,
		Uuid:      createResponse.Identity.Uuid,
		Active:    false,
	})
	require.Nil(s.T(), err)
	getResponse, err = s.nativeStub.Services.IamIdentity.Get(ctx, &identity.GetIdentityRequest{
		Namespace: namespaceName,
		Uuid:      createResponse.Identity.Uuid,
		UseCache:  false,
	})
	require.Nil(s.T(), err)
	require.False(s.T(), getResponse.Identity.Active)
}

func (s *ActiveTestSuite) TestFailsWithNotFoundErrorWhenIdentityDoesntExistInGlobalNamespace() {
	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
	defer cancel()

	createResponse, err := s.nativeStub.Services.IamIdentity.Create(ctx, &identity.CreateIdentityRequest{
		Namespace:       "",
		Name:            tools.GetRandomString(20),
		Managed:         &identity.CreateIdentityRequest_No{No: &identity.NotManagedData{}},
		InitiallyActive: true,
	})
	require.Nil(s.T(), err)
	defer s.nativeStub.Services.IamIdentity.Delete(context.Background(), &identity.DeleteIdentityRequest{Namespace: "", Uuid: createResponse.Identity.Uuid})

	_, err = s.nativeStub.Services.IamIdentity.SetActive(ctx, &identity.SetIdentityActiveRequest{
		Namespace: "",
		Uuid:      primitive.NewObjectID().Hex(),
		Active:    false,
	})
	if st, ok := status.FromError(err); ok {
		require.Equal(s.T(), codes.NotFound, st.Code())
	} else {
		require.Fail(s.T(), "Error wasnt received")
	}
}

func (s *ActiveTestSuite) TestFailsWithNotFoundErrorWhenIdentityDoesntExistInNamespace() {
	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
	defer cancel()

	namespaceName := tools.GetRandomString(20)
	_, err := s.nativeStub.Services.Namespace.Create(ctx, &namespace.CreateNamespaceRequest{
		Name:        namespaceName,
		FullName:    "",
		Description: "",
	})
	require.Nil(s.T(), err)
	defer s.nativeStub.Services.Namespace.Delete(context.Background(), &namespace.DeleteNamespaceRequest{Name: namespaceName})

	createResponse, err := s.nativeStub.Services.IamIdentity.Create(ctx, &identity.CreateIdentityRequest{
		Namespace:       namespaceName,
		Name:            tools.GetRandomString(20),
		Managed:         &identity.CreateIdentityRequest_No{No: &identity.NotManagedData{}},
		InitiallyActive: true,
	})
	require.Nil(s.T(), err)
	defer s.nativeStub.Services.IamIdentity.Delete(context.Background(), &identity.DeleteIdentityRequest{Namespace: namespaceName, Uuid: createResponse.Identity.Uuid})

	_, err = s.nativeStub.Services.IamIdentity.SetActive(ctx, &identity.SetIdentityActiveRequest{
		Namespace: namespaceName,
		Uuid:      primitive.NewObjectID().Hex(),
		Active:    false,
	})
	if st, ok := status.FromError(err); ok {
		require.Equal(s.T(), codes.NotFound, st.Code())
	} else {
		require.Fail(s.T(), "Error wasnt received")
	}
}

func (s *ActiveTestSuite) TestFailsWithNotFoundErrorWhenNamespaceDoesntExist() {
	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
	defer cancel()

	_, err := s.nativeStub.Services.IamIdentity.SetActive(ctx, &identity.SetIdentityActiveRequest{
		Namespace: tools.GetRandomString(20),
		Uuid:      primitive.NewObjectID().Hex(),
		Active:    false,
	})
	if st, ok := status.FromError(err); ok {
		require.Equal(s.T(), codes.NotFound, st.Code())
	} else {
		require.Fail(s.T(), "Error wasnt received")
	}
}
